### ISSUE TITLE
Foreach overwrites $variable with its value variable. - false positive?

### ISSUE THREAD

**Issue by lulco** [phpstan/phpstan-strict-rules#194]
I can't reproduce this issue on https://phpstan.org/try, is the rule `overwriteVariablesWithLoop` set to `true` there?

If I use foreach twice I got error on second run `Foreach overwrites $var with its value variable.` But $var is local variable here so it shouldn't be reported imho.

```
foreach ($variable as $var) {
    // do something
}

foreach ($variable as $var) {
    // do something else
}
```

**Comment by schlndh** [comment]
I found this possibly related issue: https://phpstan.org/r/ac2b69f1-1e28-4b22-a3c9-7171bae6c540

Surprisingly, changing it to `if (count($arr) > 0)`, or removing the if and adding `non-empty-array<object> $arr` makes the issue disappear.

**Comment by mvorisek** [comment]
I am facing the same issue. It is definitely a bug.

Repro: https://phpstan.org/r/c4a83ee0-e49b-4ac8-8222-17203cd19da2

The bug is present under very specific condition:
- `count($taskIds) === 0` between the same foreach loops
- with return statement between
- and the variable `$taskIds` has to be within the 1st loop

**Comment by mvorisek** [comment]
Another repro: https://phpstan.org/r/517530f9-7c8c-4ee4-aa00-2df8fd2c6c2f

Exactly the same code, but condition expression is different, same foreach code, in one, `foreach.keyOverwrite` is reported but not wanted.

**Comment by ondrejmirtes** [comment]
Is solved with: https://github.com/phpstan/phpstan-strict-rules/pull/333

Needs PHPStan 2.3.0 and phpstan-strict-rules 2.1.0 releases.
